rootchain/samuel/samuel.go

```diff
@@ -0,0 +1,580 @@
+	Prune(uint64)
+
+	// Peek returns a ready set of SAM messages, without removal
+	Peek() rootchain.VerifiedSAM
+
+	// Pop returns a ready set of SAM messages, with removal
+	Pop() rootchain.VerifiedSAM
+}
+
+// signer defines the signer interface used for
+// generating signatures
+type signer interface {
+	// Sign signs the specified data,
+	// and returns the signature and the block number at which
+	// the signature was generated
+	Sign([]byte) ([]byte, uint64, error)
+
+	// VerifySignature verifies the signature for the passed in
+	// raw data, and at the specified block number
+	VerifySignature([]byte, []byte, uint64) error
+
+	// Quorum returns the number of quorum validators
+	// for the given block number
+	Quorum(uint64) uint64
+}
+
+// transport defines the transport interface used for
+// publishing and subscribing to gossip events
+type transport interface {
+	// Publish gossips the specified SAM message
+	Publish(*proto.SAM) error
+
+	// Subscribe subscribes for incoming SAM messages
+	Subscribe(func(*proto.SAM)) error
+}
+
+// storage defines the required storage interface for SAMUEL
+// and its modules
+type storage interface {
+	// ReadLastProcessedEvent reads the last processed event data
+	ReadLastProcessedEvent(string) (string, bool)
+
+	// WriteLastProcessedEvent writes the last processed event data
+	WriteLastProcessedEvent(data string, contractAddr string) error
+}
+
+// eventData holds information on event data mapping
+type eventData struct {
+	payloadType  rootchain.PayloadType
+	eventABI     *abi.Event
+	methodABI    *abi.Method
+	localAddress types.Address
+}
+
+// SAMUEL is the module that coordinates activities with the SAMP and Event Tracker
+type SAMUEL struct {
+	eventData eventData
+	logger    hclog.Logger
+
+	eventTracker eventTracker
+	samp         samp
+	storage      storage
+	signer       signer
+	transport    transport
+}
+
+// NewSamuel creates a new SAMUEL instance
+func NewSamuel(
+	configEvent *rootchain.ConfigEvent,
+	logger hclog.Logger,
+	eventTracker eventTracker,
+	samp samp,
+	signer signer,
+	storage storage,
+	transport transport,
+) *SAMUEL {
+	return &SAMUEL{
+		logger:       logger.Named("SAMUEL"),
+		eventData:    initEventData(configEvent),
+		eventTracker: eventTracker,
+		samp:         samp,
+		signer:       signer,
+		storage:      storage,
+		transport:    transport,
+	}
+}
+
+// initEventLookupMap generates the SAMUEL event data lookup map from the
+// passed in rootchain configuration
+func initEventData(
+	configEvent *rootchain.ConfigEvent,
+) eventData {
+	return eventData{
+		payloadType:  configEvent.PayloadType,
+		eventABI:     abi.MustNewEvent(configEvent.EventABI),
+		methodABI:    abi.MustNewABI(configEvent.MethodABI).GetMethod(configEvent.MethodName),
+		localAddress: types.StringToAddress(configEvent.LocalAddress),
+	}
+}
+
+// Start starts the SAMUEL module
+func (s *SAMUEL) Start() error {
+	// Start the event loop for the tracker
+	s.startEventLoop()
```

Comment:
If the below initialization processes fail, goroutine will be left, I think. Is it impossible to move `startEventLoop` to the bottom of the method?

Reply:
This is a good catch 👍 

Moved it, it doesn't make sense to listen for events if the event tracker start fails:

[2f29004](https://github.com/0xPolygon/polygon-edge/pull/717/commits/2f2900400f619dd610e4bc76ad9aae28a0fa6f87)